overlord/snapstate/snapstate.go

```diff
@@ -427,6 +428,41 @@ func CheckChangeConflict(st *state.State, snapName string, checkConflictPredicat
 	return nil
 }
 
+func contentAttr(attrs map[string]interface{}) string {
+	// we always have a "content" attr in plug/slots of interface
+	// type "content"
+	s, _ := attrs["content"].(string)
+	return s
+}
+
+func contentIfaceAvailable(st *state.State, contentTag string) bool {
+	repo := ifacerepo.Get(st)
+	for _, slot := range repo.AllSlots("content") {
+		if contentAttr(slot.Attrs) == contentTag {
+			return true
+		}
+	}
+	return false
+}
+
+// defaultContentPlugProviders takes a snap.Info and returns what
+// default providers there are.
+func defaultContentPlugProviders(st *state.State, info *snap.Info) []string {
+	out := []string{}
+	for _, plug := range info.Plugs {
+		if plug.Interface == "content" {
+			if !contentIfaceAvailable(st, contentAttr(plug.Attrs)) {
+				dprovider, ok := plug.Attrs["default-provider"].(string)
```

Comment:
I think we should make sure this attribute is sane in Sanitize method of the plug (after you've read the snap yaml the plugs/slots are already sanitized).

Reply:
I think this makes sense but unfortunately we have apparently already snaps in the wild that have strange default provider names. So we need to think how we sanitize (warn ideally) without breaking snaps that exist and used to work.